File: batch_process.m
function batch_process()
    %BATCH_PROCESS Processes all images in the images folder
    %
    % This script processes all .bmp images in the images directory
    % using the LIME enhancement algorithm.
    %
    % Example:
    %   batch_process()
    
    % Get script directory
    scriptDir = fileparts(mfilename('fullpath'));
    
    % Load configuration
    cfg = config();
    
    % Create output directory if it doesn't exist
    if cfg.saveResults && ~exist(cfg.outputDir, 'dir')
        mkdir(cfg.outputDir);
    end
    
    % Get all .bmp files in images directory
    imagesDir = fullfile(scriptDir, 'images');
    imageFiles = dir(fullfile(imagesDir, '*.bmp'));
    
    if isempty(imageFiles)
        fprintf('No .bmp images found in %s\n', imagesDir);
        return;
    end
    
    fprintf('Found %d images to process\n\n', length(imageFiles));
    
    % Create LIME enhancer with configured parameters
    enhancer = LIME('alpha', cfg.alpha, ...
                    'gamma', cfg.gamma, ...
                    'rho', cfg.rho, ...
                    'mu', cfg.mu, ...
                    'numIterations', cfg.numIterations);
    
    % Process each image
    for i = 1:length(imageFiles)
        imagePath = fullfile(imagesDir, imageFiles(i).name);
        fprintf('[%d/%d] Processing: %s\n', i, length(imageFiles), imageFiles(i).name);
        
        try
            % Load input image
            inputImage = imread(imagePath);
            
            % Enhance image
            [enhancedImage, results] = enhancer.enhance(inputImage);
            
            % Save results if requested
            if cfg.saveResults
                [~, baseName, ~] = fileparts(imageFiles(i).name);
                
                % Save initial enhanced image
                initialPath = fullfile(cfg.outputDir, [baseName, '_initial.png']);
                imwrite(results.initialEnhanced, initialPath);
                
                % Save refined enhanced image
                refinedPath = fullfile(cfg.outputDir, [baseName, '_refined.png']);
                imwrite(results.refinedEnhanced, refinedPath);
                
                % Save final enhanced image
                finalPath = fullfile(cfg.outputDir, [baseName, '_enhanced.png']);
                imwrite(enhancedImage, finalPath);
                
                fprintf('  ✓ Saved: %s_*.png\n', baseName);
            end
            
        catch ME
            fprintf('  ✗ Error processing %s: %s\n', imageFiles(i).name, ME.message);
        end
        
        fprintf('\n');
    end
    
    fprintf('Batch processing complete! Processed %d images.\n', length(imageFiles));
end

File: python/batch_process.py
"""
Batch processing script for LIME image enhancement.

This script processes all images in the images folder using the LIME
enhancement algorithm.
"""
from pathlib import Path
import numpy as np
from PIL import Image

from config import get_config
from lime import LIME


def batch_process():
    """Processes all images in the images folder."""
    # Get script directory
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    
    # Load configuration
    cfg = get_config()
    
    # Create output directory if it doesn't exist
    if cfg.save_results:
        output_path = project_root / cfg.output_dir
        output_path.mkdir(parents=True, exist_ok=True)
    
    # Get all image files in images directory
    images_dir = project_root / 'images'
    
    # Support multiple image formats
    image_extensions = ['.bmp', '.png', '.jpg', '.jpeg']
    image_files = []
    for ext in image_extensions:
        image_files.extend(list(images_dir.glob(f'*{ext}')))
        image_files.extend(list(images_dir.glob(f'*{ext.upper()}')))
    
    if not image_files:
        print(f'No images found in {images_dir}')
        return
    
    print(f'Found {len(image_files)} images to process\n')
    
    # Create LIME enhancer with configured parameters
    enhancer = LIME(alpha=cfg.alpha,
                    gamma=cfg.gamma,
                    rho=cfg.rho,
                    mu=cfg.mu,
                    num_iterations=cfg.num_iterations)
    
    # Process each image
    for i, image_path in enumerate(image_files, 1):
        print(f'[{i}/{len(image_files)}] Processing: {image_path.name}')
        
        try:
            # Load input image
            input_image = np.array(Image.open(image_path))
            
            # Enhance image
            enhanced_image, results = enhancer.enhance(input_image)
            
            # Save results if requested
            if cfg.save_results:
                base_name = image_path.stem
                output_path = project_root / cfg.output_dir
                
                # Save initial enhanced image
                initial_path = output_path / f'{base_name}_initial.png'
                initial_img = Image.fromarray((np.clip(results['initial_enhanced'], 0, 1) * 255).astype(np.uint8))
                initial_img.save(initial_path)
                
                # Save refined enhanced image
                refined_path = output_path / f'{base_name}_refined.png'
                refined_img = Image.fromarray((np.clip(results['refined_enhanced'], 0, 1) * 255).astype(np.uint8))
                refined_img.save(refined_path)
                
                # Save final enhanced image
                final_path = output_path / f'{base_name}_enhanced.png'
                final_img = Image.fromarray((np.clip(enhanced_image, 0, 1) * 255).astype(np.uint8))
                final_img.save(final_path)
                
                print(f'  ✓ Saved: {base_name}_*.png')
            
        except Exception as e:
            print(f'  ✗ Error processing {image_path.name}: {str(e)}')
        
        print()
    
    print(f'Batch processing complete! Processed {len(image_files)} images.')


if __name__ == '__main__':
    batch_process()

